app/uk/gov/hmrc/gform/obligation/ObligationService.scala

```diff
@@ -17,30 +17,131 @@
+                          i =>
+                            authService.evaluateSubmissionReference(
+                              i.idNumber,
+                              retrievals,
+                              formTemplate,
+                              FormDataHelpers.formDataMap(form.formData),
+                              form.envelopeId))
+      output <- gformConnector
+                 .getAllTaxPeriods(
+                   NonEmptyList.fromList(
+                     listOfIdNumbers
+                       .zip(hmrcTaxPeriodIdentifiers)
+                       .map(a => HmrcTaxPeriod(a._2.idType, TextExpression(Constant(a._1)), a._2.regimeType))) match {
+                     case Some(x) => x
+                   })
+                 .map(i =>
+                   i.flatMap(j => j.obligation.obligations.flatMap(h => makeAllInfoList(j.id, h.obligationDetails))))
+    } yield output
+  }
+
+  def makeAllInfoList(id: HmrcTaxPeriod, obligation: NonEmptyList[ObligationDetail]) =
+    obligation.map(i =>
+      TaxPeriodInformation(id, i.inboundCorrespondenceFromDate, i.inboundCorrespondenceToDate, i.periodKey))
+
+  def lookupIfPossible(
+    form: Form,
+    formTemplate: FormTemplate,
+    authService: AuthService,
+    retrievals: MaterialisedRetrievals)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Form] = {
     val hmrcTaxPeriodIdentifiers = formTemplate.expandFormTemplateFull.allFCs.collect {
       case IsHmrcTaxPeriod(el) => el
     }
-    val futureListOfTaxPeriodDes =
-      gformConnector
-        .getAllTaxPeriods(hmrcTaxPeriodIdentifiers)
-        .map(i => i.flatMap(j => j.obligation.obligations.flatMap(h => makeMap(j.id, h.obligationDetails))))
-    futureListOfTaxPeriodDes.map(i => i.toMap)
+
+    hmrcTaxPeriodIdentifiers match {
+      case a :: _ =>
+        for {
+          currentIdNumber <- authService.evaluateSubmissionReference(
+                              a.idNumber,
+                              retrievals,
+                              formTemplate,
+                              FormDataHelpers.formDataMap(form.formData),
+                              form.envelopeId
+                            )
+          output <- form.obligations match {
+                     case RetrievedObligations(x) => {
+                       for {
+                         idNumbers <- Future.traverse(x.toList)(
+                                       i =>
+                                         authService.evaluateSubmissionReference(
+                                           i.hmrcTaxPeriod.idNumber,
+                                           retrievals,
+                                           formTemplate,
+                                           FormDataHelpers.formDataMap(form.formData),
+                                           form.envelopeId))
+                         output <- {
+                           ifStatement(
```

Comment:
This `ifStatement` would be much better when defined as local method in `lookupIfPossible`. Than all it need to take is `condition: Boolean` as a parameter. All the rest is available in scope.